File: batch_processing.py
import time
import random

# Function to generate fake requests
def generate_fake_requests(num_requests):
    requests = [i for i in range(num_requests)]
    random.shuffle(requests)
    return requests

# Server class to handle batch processing
class Server:
    def __init__(self, batch_size=50):
        self.batch_size = batch_size
        self.requests = []

    def collect_requests(self, new_requests):
        self.requests.extend(new_requests)

    def process_batches(self):
        while self.requests:
            batch = self.requests[:self.batch_size]
            self.requests = self.requests[self.batch_size:]
            self.process_batch(batch)

    def process_batch(self, batch):
        print(f"Processing batch: {batch}")
        time.sleep(2)  # Simulating 2 seconds of processing time
        print(f"Batch processed: {batch}")

# Main function
def main():
    num_requests = 100
    fake_requests = generate_fake_requests(num_requests)

    server = Server(batch_size=50)

    # Collect requests in batches of 50
    for i in range(0, num_requests, 50):
        server.collect_requests(fake_requests[i:i+50])
        print(f"Collected batch {i//50 + 1} of 4 batches.")

    # Process the batches
    server.process_batches()

if __name__ == "__main__":
    main()

File: batch_processing.js
// Function to generate fake requests
function generateFakeRequests(numRequests) {
  // Create an array of sequential integers from 0 to (numRequests - 1)
  const requests = Array.from({ length: numRequests }, (_, i) => i);

  // Shuffle the array of requests to create a random order
  shuffleArray(requests);

  // Return the array of shuffled requests
  return requests;
}

// Function to shuffle an array
function shuffleArray(array) {
  // Loop through the array from the last index to the second index
  for (let i = array.length - 1; i > 0; i--) {
    // Generate a random index between 0 and i (inclusive)
    const j = Math.floor(Math.random() * (i + 1));

    // Swap the elements at indices i and j
    [array[i], array[j]] = [array[j], array[i]];
  }
}

// Server class to handle batch processing
class Server {
  constructor(batchSize = 50) {
    // Set the batch size for processing requests
    this.batchSize = batchSize;

    // Initialize an empty array to store the received requests
    this.requests = [];
  }

  // Function to collect new requests and add them to the requests array
  collectRequests(newRequests) {
    this.requests.push(...newRequests);
  }

  // Function to process all the batches of requests
  async processBatches() {
    // Loop until there are no more requests in the array
    while (this.requests.length) {
      // Extract a batch of requests (up to this.batchSize)
      const batch = this.requests.splice(0, this.batchSize);

      // Process the current batch
      await this.processBatch(batch);
    }
  }

  // Function to process a single batch of requests
  async processBatch(batch) {
    // Log the batch being processed
    console.log(`Processing batch: [${batch}]`);

    // Create a new Promise that will resolve after a 2-second delay
    return new Promise((resolve) => {
      setTimeout(() => {
        // Log the batch after it has been processed
        console.log(`Batch processed: [${batch}]`);

        // Resolve the Promise to indicate that the batch has been processed
        resolve();
      }, 2000);
    });
  }
}

// Main function
async function main() {
  // Generate 200 fake requests
  const numRequests = 200;
  const fakeRequests = generateFakeRequests(numRequests);

  // Create a new Server instance with a batch size of 50
  const server = new Server(batchSize = 50);

  // Collect the requests in batches of 50
  for (let i = 0; i < numRequests; i += 50) {
    // Get the next batch of 50 requests
    const batch = fakeRequests.slice(i, i + 50);

    // Add the batch to the server's request collection
    server.collectRequests(batch);

    // Log a message indicating that a batch has been collected
    console.log(`Collected batch ${Math.floor(i / 50) + 1} of 4 batches.`);
  }

  // Process all the batches of requests
  await server.processBatches();
}

// Call the main function
main();